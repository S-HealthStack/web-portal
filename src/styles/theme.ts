import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
