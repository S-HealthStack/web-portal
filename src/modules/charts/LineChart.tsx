import React, { useCallback, useRef, useMemo, useState, useEffect } from 'react';
import styled, { useTheme } from 'styled-components';
import * as d3 from 'd3';

import { SpecColorType } from 'src/styles/theme';
import Tooltip from 'src/common/components/Tooltip';
import { format } from 'src/common/utils/datetime';
import {
  getFocusWidth,
  getFocusHeight,
  getDefaultXRange,
  getDefaultYRange,
  getTooltipPosition,
  MARGIN_FOCUS,
  Container,
  TooltipContent,
  formatTimeAxisWithNames,
  DOTS_CONTAINER_CLASS_NAME,
  TooltipProps,
  StyledSvg,
  TooltipNameBlock,
  TooltipDataBlock,
  TooltipColorPoint,
} from './common-helpers';
import { Line, Area, Dot, DotDataItem, XAxis, YAxis } from './common-components';

export interface DataItem {
  // TODO there should be participant id
  name: string;
  ts: number;
  value: number;
  min: number;
  max: number;
  highlighted: boolean;
  lastSync: number;
  color: SpecColorType;
}

type Props = {
  data: DataItem[];
  xDomain: number[];
  width: number;
  height: number;
  showTrendLines?: boolean;
  hiddenDataLines: string[];
  onDotClick: (_: React.MouseEvent<SVGPathElement, MouseEvent>, d: DataItem) => void; // TODO we need participant id from data, that's why unused attributes here
};

const getDotId = (name: string, ts: number) => `dot-${name}-${ts}`;

const SvgContainer = styled(StyledSvg)`
  .focus {
    cursor: default;
  }
`;

const LineChart = ({
  width,
  height,
  data,
  xDomain,
  showTrendLines,
  hiddenDataLines,
  onDotClick,
}: Props) => {
  const svgRef = useRef<SVGSVGElement>(null);
  const containerRef = useRef<HTMLDivElement>(null);

  const theme = useTheme();

  const [tooltipProps, setTooltipProps] = useState<TooltipProps>(null);
  const [enteredDotId, setEnteredDotId] = useState<string>('');

  const chartData = useMemo(() => {
    const groupedData = d3.group(data, (d) => d.name);
    const groupNames = d3.map(groupedData, (d) => d[0]);
    const sortedChartData = d3.map(groupedData, (d) => {
      const ds = [...d[1]];
      ds.sort((item1, item2) => item1.ts - item2.ts);
      return ds;
    });

    const colorScale = d3.scaleOrdinal(
      groupNames,
      d3.map(groupedData, (d) => theme.colors[d[1][0].color])
    );

    return { sortedChartData, colorScale };
  }, [data, theme.colors]);

  const yStartDomain = useMemo(() => {
    const minData = d3.min(data, (d) => (d.min ? +d.min : 0)) || 0;
    const maxData = d3.max(data, (d) => (d.max ? +d.max : 0)) || 0;

    return [Math.floor(minData / 10) * 10, Math.ceil(maxData / 10) * 10];
  }, [data]);

  const xScale = useMemo(
    () => d3.scaleTime().domain(xDomain).range(getDefaultXRange(width)),
    [xDomain, width]
  );

  const yScale = useMemo(
    () => d3.scaleLinear().domain(yStartDomain).nice().range(getDefaultYRange(height)),
    [height, yStartDomain]
  );

  const showDot = useCallback(
    (d: DataItem) => !hiddenDataLines?.includes(d.name) && d.highlighted,
    [hiddenDataLines]
  );

  const onDotMouseEnter = useCallback(
    (event: React.MouseEvent<SVGPathElement, MouseEvent>, enteredDot: DotDataItem) => {
      if (!svgRef.current) {
        return;
      }

      const svgRect = svgRef.current.getBoundingClientRect();
      const dotNode = d3
        .select(svgRef.current)
        .select<Element>(`#${enteredDot.id}`)
        .node()
        ?.getBoundingClientRect();
      const color = chartData.colorScale(enteredDot.name);

      const tooltipContent = (
        <TooltipContent>
          <TooltipNameBlock>
            <TooltipColorPoint color={color} />
            <p>{`${enteredDot.name}:` /* TODO: replace name by participant id */}</p>
          </TooltipNameBlock>
          <TooltipDataBlock>
            <p>{Math.round(enteredDot.y)} bpm</p>
            <p>{format(enteredDot.lastSync, 'hh:mm a')}</p>
          </TooltipDataBlock>
        </TooltipContent>
      );

      setTooltipProps({
        content: tooltipContent,
        point: dotNode
          ? [
              dotNode.x - svgRect.left + dotNode.width / 2,
              dotNode.y - svgRect.top + dotNode.height / 2,
            ]
          : [event.pageX - svgRect.left, event.pageY - svgRect.top],
        position: getTooltipPosition(event.pageX, event.pageY, svgRect),
      });

      if (enteredDot.id) {
        setEnteredDotId(enteredDot.id);
      }
    },
    [chartData]
  );

  const resetSelectedDot = useCallback(() => {
    setTooltipProps(null);
    setEnteredDotId('');
  }, []);

  const getDotOpacity = useCallback(
    (d: DataItem, dotId: string) => {
      if (!showDot(d)) {
        return 0;
      }

      if (!tooltipProps || enteredDotId === dotId) {
        return 1;
      }

      return 0.5;
    },
    [enteredDotId, showDot, tooltipProps]
  );

  const xAxis = useMemo(
    () => (
      <XAxis
        key="x-axis"
        xScale={xScale}
        tickSize={getFocusHeight(height)}
        yOffset={height - MARGIN_FOCUS.bottom}
        yTickOffset={20}
        ticks={6}
        tickFormatFn={formatTimeAxisWithNames}
        removeDomain
      />
    ),
    [height, xScale]
  );

  const yAxis = useMemo(
    () => (
      <YAxis
        key="y-axis"
        yScale={yScale}
        tickSize={getFocusWidth(width)}
        xOffset={MARGIN_FOCUS.left}
        xTickOffset={-20}
        ticks={4}
        removeDomain
      />
    ),
    [width, yScale]
  );

  const lineItems = useMemo(
    () =>
      chartData.sortedChartData.map((dataItem) => {
        const id = `line-${dataItem[0].name}`;
        return (
          <Line
            key={id}
            id={id}
            xScale={xScale}
            yScale={yScale}
            data={dataItem.map((d) => ({ x: d.ts, y: d.value }))}
            color={chartData.colorScale(dataItem[0].name)}
            width={4}
            strokeOpacity={
              // eslint-disable-next-line no-nested-ternary
              !hiddenDataLines?.includes(dataItem[0].name) ? (tooltipProps ? 0.5 : 1) : 0
            }
          />
        );
      }),
    [chartData, hiddenDataLines, tooltipProps, xScale, yScale]
  );

  const areaItems = useMemo(
    () =>
      chartData.sortedChartData.map((dataItem) => {
        const id = `area-${dataItem[0].name}`;
        return (
          <Area
            key={id}
            id={id}
            xScale={xScale}
            yScale={yScale}
            data={dataItem.map((d) => ({ x: d.ts, y0: d.min, y1: d.max }))}
            color={chartData.colorScale(dataItem[0].name)}
            visible={!hiddenDataLines?.includes(dataItem[0].name) && showTrendLines}
          />
        );
      }),
    [chartData, hiddenDataLines, showTrendLines, xScale, yScale]
  );

  const dotItems = useMemo(
    () => (
      <g key="dots" className={DOTS_CONTAINER_CLASS_NAME}>
        {chartData.sortedChartData.map((dataItem) =>
          dataItem.map(
            (d) =>
              d.highlighted && (
                <Dot
                  key={getDotId(d.name, d.ts)}
                  id={getDotId(d.name, d.ts)}
                  xScale={xScale}
                  yScale={yScale}
                  data={{ x: d.ts, y: d.value, lastSync: d.lastSync, name: d.name }}
                  color={theme.colors.onSurface}
                  fillOpacity={getDotOpacity(d, `dot-${d.name}-${d.ts}`)}
                  onMouseEnter={onDotMouseEnter}
                  onClick={(event) => onDotClick(event, d)}
                />
              )
          )
        )}
      </g>
    ),
    [
      chartData.sortedChartData,
      getDotOpacity,
      onDotClick,
      onDotMouseEnter,
      theme.colors.onSurface,
      xScale,
      yScale,
    ]
  );

  useEffect(() => {
    d3.select(svgRef.current)
      .select<SVGGElement>('.focus')
      .on('mouseenter', () => tooltipProps && resetSelectedDot());
  });

  if (!width || !height || !data.length) {
    return null;
  }

  return (
    <Container width={width} height={height} ref={containerRef}>
      <SvgContainer
        ref={svgRef}
        focusWidth={getFocusWidth(width)}
        viewBox={`0, 0, ${width}, ${height}`}
        $contextVisible={false}
      >
        {xAxis}
        {yAxis}
        {lineItems}
        {areaItems}
        <g key="focus" className="focus">
          <rect
            transform={`translate(${MARGIN_FOCUS.left}, ${MARGIN_FOCUS.top})`}
            width={getFocusWidth(width)}
            height={getFocusHeight(height)}
            fill="transparent"
          />
        </g>
        {dotItems}
      </SvgContainer>
      <Tooltip
        static
        content={tooltipProps?.content}
        point={tooltipProps?.point}
        show={!!tooltipProps}
        position={tooltipProps?.position}
        arrow
      />
    </Container>
  );
};

export default LineChart;
