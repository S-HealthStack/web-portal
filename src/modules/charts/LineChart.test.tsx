import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { act, findByTestId, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components/';
import { userEvent } from '@storybook/testing-library';
import theme from 'src/styles/theme';
import LineChart from 'src/modules/charts/LineChart';

describe('LineChart', () => {
  it('should render', async () => {
    const onDotClick = jest.fn();

    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <LineChart
          data={[
            {
              name: 'test',
              ts: 0,
              value: 10,
              min: 5,
              max: 20,
              highlighted: true,
              lastSync: 0,
              color: 'primary',
            },
            {
              name: 'test',
              ts: 1,
              value: 11,
              min: 5,
              max: 20,
              highlighted: false,
              lastSync: 0,
              color: 'primary',
            },
          ]}
          width={1000}
          height={1000}
          hiddenDataLines={['none']}
          onDotClick={onDotClick}
          xDomain={[0, 100]}
          showTrendLines
        />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const dot = await findByTestId(baseElement, 'dot-test-0');
    expect(dot).toBeInTheDocument();

    act(() => {
      userEvent.hover(dot);
    });

    act(() => {
      userEvent.click(dot);
    });

    expect(onDotClick).toHaveBeenCalled();

    act(() => {
      userEvent.hover(baseElement.querySelector('.focus') as HTMLElement);
    });
  });

  it('[NEGATIVE] should render with empty data', () => {
    const { baseElement } = render(
      <ThemeProvider theme={theme}>
        <LineChart
          data={[]}
          width={1000}
          height={1000}
          hiddenDataLines={['none']}
          onDotClick={() => {}}
          xDomain={[0, 1]}
        />
      </ThemeProvider>
    );

    expect(baseElement).toBeDefined();
  });
});
