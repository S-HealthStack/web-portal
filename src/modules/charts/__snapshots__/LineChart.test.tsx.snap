// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart should render 1`] = `
.c4 {
  display: inline-block;
  position: relative;
}

.c0 .xAxis .tick line {
  opacity: 0;
}

.c0 .xAxis .tick:first-child text {
  text-anchor: start;
}

.c0 .xAxis .tick:last-child text {
  text-anchor: end;
}

.c0 .yAxis .tick line {
  stroke: #F0F0F0;
}

.c0 .context {
  opacity: 0;
  -webkit-transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: opacity 0.3s cubic-bezier(0.45,0.05,0,1);
}

.c0 .context .brushX .overlay {
  fill: #4475E3;
  fill-opacity: 0.1;
  pointer-events: none;
  height: 0.25rem;
}

.c0 .context .brushX .selection {
  fill: #4475E3;
  fill-opacity: 1;
  height: 1.375rem;
  fill: url(#horizontalGradient);
}

.c0 .context .brushX .handle {
  stroke-width: 0.0625rem;
  stroke: #4475E3;
  stroke-opacity: 0.3;
  fill: #FFFFFF;
  height: 1.375rem;
  width: 0.375rem;
}

.c0 .context .brushY .overlay {
  fill: #4475E3;
  fill-opacity: 0.1;
  pointer-events: none;
  width: 0.25rem;
}

.c0 .context .brushY .selection {
  fill: #4475E3;
  fill-opacity: 1;
  width: 1.375rem;
  fill: url(#verticalGradient);
}

.c0 .context .brushY .handle {
  stroke-width: 0.0625rem;
  stroke: #4475E3;
  stroke-opacity: 0.3;
  fill: #FFFFFF;
  height: 0.375rem;
  width: 1.375rem;
}

.c0 .focus {
  cursor: move;
}

.c0 .line {
  -webkit-clip-path: url(#clip);
  clip-path: url(#clip);
  -webkit-transition: stroke-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: stroke-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
}

.c0 .area {
  -webkit-clip-path: url(#clip);
  clip-path: url(#clip);
  -webkit-transition: fill-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: fill-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
}

.c0 .dots {
  -webkit-clip-path: url(#clip);
  clip-path: url(#clip);
}

.c0 .dots .dot {
  -webkit-transition: fill-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
  transition: fill-opacity 0.3s cubic-bezier(0.45,0.05,0,1);
}

.c2 > .tick > text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
}

.c3 > .tick > text {
  font-family: Inter,sans-serif;
  font-size: 0.625rem;
  line-height: 0.9375rem;
  font-weight: 400;
  font-style: normal;
}

.c1 .focus {
  cursor: default;
}

<body>
  <div>
    <div
      class=""
      height="1000"
      style="width: 62.5rem; height: 62.5rem; position: relative;"
      width="1000"
    >
      <svg
        class="c0 c1"
        viewBox="0, 0, 1000, 1000"
      >
        <g
          class="c2 xAxis"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="middle"
          transform="translate(0, 978)"
        >
          <g
            class="tick"
            opacity="1"
            transform="translate(24.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-973"
            />
            <text
              dy="0em"
              fill="currentColor"
              y="20"
            >
              MIDNIGHT
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(217.3,0)"
          >
            <line
              stroke="currentColor"
              y2="-973"
            />
            <text
              dy="0em"
              fill="currentColor"
              y="20"
            >
              MIDNIGHT
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(410.1,0)"
          >
            <line
              stroke="currentColor"
              y2="-973"
            />
            <text
              dy="0em"
              fill="currentColor"
              y="20"
            >
              MIDNIGHT
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(602.9,0)"
          >
            <line
              stroke="currentColor"
              y2="-973"
            />
            <text
              dy="0em"
              fill="currentColor"
              y="20"
            >
              MIDNIGHT
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(795.7,0)"
          >
            <line
              stroke="currentColor"
              y2="-973"
            />
            <text
              dy="0em"
              fill="currentColor"
              y="20"
            >
              MIDNIGHT
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(988.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-973"
            />
            <text
              dy="0em"
              fill="currentColor"
              y="20"
            >
              MIDNIGHT
            </text>
          </g>
        </g>
        <g
          class="c3 yAxis"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="start"
          transform="translate(24,0)"
        >
          <g
            class="tick"
            opacity="1"
            transform="translate(0,978.5)"
          >
            <line
              stroke="currentColor"
              x2="964"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-20"
            >
              0
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,735.25)"
          >
            <line
              stroke="currentColor"
              x2="964"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-20"
            >
              5
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,492)"
          >
            <line
              stroke="currentColor"
              x2="964"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-20"
            >
              10
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,248.75)"
          >
            <line
              stroke="currentColor"
              x2="964"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-20"
            >
              15
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,5.5)"
          >
            <line
              stroke="currentColor"
              x2="964"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-20"
            >
              20
            </text>
          </g>
        </g>
        <path
          class="line"
          d="M24,491.5L33.64,442.84999999999997"
          data-testid="line-test"
          fill="none"
          id="line-test"
          stroke="#4475E3"
          stroke-opacity="1"
          stroke-width="2"
        />
        <path
          class="area"
          d="M24,5L33.64,5L33.64,734.75L24,734.75Z"
          data-testid="area-test"
          fill="#4475E3"
          fill-opacity="0.1"
          id="area-test"
        />
        <g
          class="focus"
        >
          <rect
            fill="transparent"
            height="973"
            transform="translate(24, 5)"
            width="964"
          />
        </g>
        <g
          class="dots"
        >
          <circle
            class="dot"
            cursor="pointer"
            data-testid="dot-test-0"
            fill="#2D2D2D"
            fill-opacity="1"
            id="dot-test-0"
            pointer-events="all"
            r="4"
            transform="translate(24,491.5)"
          />
        </g>
      </svg>
      <span
        class="c4"
        data-testid="tooltip-container"
      />
    </div>
  </div>
</body>
`;
